blogposts/2021/optimizing-a-code-intelligence-commit-graph.md

```diff
@@ -0,0 +1,699 @@
+
+## Hitting a moving target
+
+While we were busy preparing a patch release to restore code intelligence to our customer, they had upgraded to the next version to fix an unrelated error they were experiencing in Sourcegraph's search code.
+
+This made things **far** worse. In Sourcegraph 3.22, we moved the code that calculated the commit graph from the precise-code-intel-worker into the frontend in an effort to consolidate background and period processes into the same package. Our first stab at memory reduction wasn't quite _enough_, and now the frontend pods were taking turns using all available memory and crashing.
+
+This made their entire instance unstable, and this series of events escalated us from a priority zero event to a [priority now](https://devblogs.microsoft.com/oldnewthing/20081121-00/?p=20123) event.
+
+We attacked the problem again, starting out again by tackling some low-hanging fruit by ignoring large gobs of data we ended up not needing. When we pull back the commit graph for a repository, it's unlikely that we need the _entire_ commit graph. There's little sense in filling out the visibility of the long tale of historic commits, especially as its distance to the oldest LSIF upload grows over time. Now, we [entirely ignore](https://github.com/sourcegraph/sourcegraph/pull/16140) the portion of the commit graph that existed _before_ the oldest known LSIF upload fro that repository.
+
+In the same spirit, we can remove the explicit step of topologically sorting the commit graph in the application by instead [sorting it via the git command](https://github.com/sourcegraph/sourcegraph/pull/16270). This is basically the same win as replacing a `sort` call in your webapp with an `ORDER BY` clause in your SQL query. This further reduces the required memory as we're no longer taking large amounts of stack space to traverse long chains of commits in the graph.
+
+Our [last change](https://github.com/sourcegraph/sourcegraph/pull/16368), however, was the real heavy hitter. So far, we've been storing our commit graph in a map from commits to the set of uploads visible from that commit. When the number number of distinct roots is large, these lists are also quite large. Most notably, this makes the merge operation between two lists quite expensive in terms of both CPU (the original merge procedure compares each pair of elements from the list in a trivial but quadratic nested loop) and memory (merging two lists creates a third, all of which are live at the same time).
+
+At a slight expense of memory, we can greatly optimize CPU time by instead by slightly altering this mapping. Note that looking in a single direction, each commit can see at most one upload for a particular indexer and root. Then, instead of storing a flat list of visible uploads per commit, we can store a map from the indexer/root values to the visible upload with those properties. Merge operations now become linear in the size of the input lists.
```

Comment:
```suggestion
At a slight expense of memory, we can greatly optimize CPU time by instead slightly altering this mapping. Note that looking in a single direction, each commit can see at most one upload for a particular indexer and root. Then, instead of storing a flat list of visible uploads per commit, we can store a map from the indexer/root values to the visible upload with those properties. Merge operations now become linear in the size of the input lists.
```